src/hotspot/share/gc/shenandoah/shenandoahAgeCensus.cpp

```diff
@@ -0,0 +1,271 @@
+      _local_age_table[i]->clear();   // clear for next census
+      // Merge noise stats
+      CENSUS_NOISE(_global_noise[_epoch].merge(_local_noise[i]);)
+      CENSUS_NOISE(_local_noise[i].clear();)
+    }
+
+    compute_tenuring_threshold();
+  }
+}
+
+
+// Reset the epoch for the global age tables,
+// clearing all history.
+// TBD: do this for noise & local tables too?
+void ShenandoahAgeCensus::reset_epoch() {
+  assert(_epoch < MAX_SNAPSHOTS, "Out of bounds");
+  for (uint i = 0; i < MAX_SNAPSHOTS; i++) {
+    _global_age_table[i]->clear();
+  }
+  _epoch = MAX_SNAPSHOTS;
+  assert(_epoch < MAX_SNAPSHOTS, "Error");
+}
+
+// Ingest a population vector into the global age table for
+// the current epoch, merging it with any current data already
+// present.
+void ShenandoahAgeCensus::ingest(AgeTable* population_vector) {
+  _global_age_table[_epoch]->merge(population_vector);
+}
+
+void ShenandoahAgeCensus::compute_tenuring_threshold() {
+  if (!GenShenAdaptiveTenuring) {
+    _tenuring_threshold[_epoch] = InitialTenuringThreshold;
+  } else {
+    uint tt = compute_tenuring_threshold_work();
+    assert(tt <= MAX_COHORTS, "Out of bounds");
+    _tenuring_threshold[_epoch] = tt;
+  }
+  print();
+  log_trace(gc, age)("New tenuring threshold " UINTX_FORMAT
+    " (min " UINTX_FORMAT ", max " UINTX_FORMAT")",
+    (uintx) _tenuring_threshold[_epoch], GenShenMinTenuringThreshold, GenShenMaxTenuringThreshold);
+}
+
+uint ShenandoahAgeCensus::compute_tenuring_threshold_work() {
+  // Starting with the largest non-zero population by age cohort
+  // and working down in age of cohorts,find the lowest age such
+  // that all higher ages have a mortality rate that is below a
+  // pre-specified threshold. We consider this to be the adaptive
+  // tenuring age to be used for the next cohort.
+  // Results are clamped between user-specified mix & max guardrails,
+  // so we ignore any cohorts outside [min,max].
+
+  // Current and previous epoch in ring
+  const uint cur_epoch = _epoch;
+  const uint prev_epoch = cur_epoch > 0  ? cur_epoch - 1 : markWord::max_age;
+  uint tenuring_threshold = GenShenMaxTenuringThreshold;
+
+  // Current and previous population vectors in ring
+  const AgeTable* cur_pv = _global_age_table[cur_epoch];
+  const AgeTable* prev_pv = _global_age_table[prev_epoch];
+  for (uint i = GenShenMaxTenuringThreshold - 1; i >= MAX2((uint)GenShenMinTenuringThreshold - 1, (uint)1); i--) {
+    assert(i > 0, "Error");
+    // Population & mortality rate of current cohort
+    const size_t cur_pop = cur_pv->sizes[i];
+    const size_t prev_pop = prev_pv->sizes[i-1];
+    const double mr = mortality_rate(prev_pop, cur_pop);
+    // We ignore any cohorts that had a very low population count, or
+    // that have a lower mortality rate than we care to age in young; these
+    // cohorts are considered eligible for tenuring when all older
+    // cohorts are.
+    if (i > 1 && (prev_pop < GenShenTenuringCohortPopulationThreshold ||
```

Comment:
Consider getting a more accurate population for cohort 0 (by adding the population in each region above TAMS), so that we may consider the case of `i = 1` also above (i.e. `i >=1`). This would potentially allo tenuring threshold to go as low as 1 in some cases. This choice will need to be carefully considered.

Reply:
Done. Default ceiling is set by min age which defaults to 1.